blog/2025-06-15-temporal-impl-1.md

```diff
@@ -0,0 +1,439 @@
+---
+layout: post
+tags: [post]
+title:
+  Implementing Temporal, the new date/time API for JavaScript (and
+  Rust!)
+metadata: ["temporal", "temporal_rs", "boa", "date/time"]
+description:
+  A blog post about the temporal_rs Rust crate that implements
+  JavaScript's Temporal API and how temporal_rs supports implementing
+  Temporal in JavaScript engines.
+authors: boa-dev
+---
+
+This will be a series of posts primarily about implementing a new
+JavaScript feature in Rust, specifically the new date/time built-in:
+Temporal. We'll be going over the general implementation of Temporal in
+Boa as well as the crates supporting that implementation.
+
+Why should you care? Well we are not only implementing it for
+JavaScript, but Rust as well ... more on that in a bit.
+
+First, an aside!
+
+## What even is Temporal?
+
+Temporal is a modern API for handling date/time in a calendar and time
+zone aware manner that includes nine objects with over 200+ methods.
+
+In JavaScript, Temporal is a global built-in namespace object that
+includes each of these nine built-ins:
+
+- `Temporal.Now`
+- `Temporal.PlainDate`
+- `Temporal.PlainTime`
+- `Temporal.PlainDateTime`
+- `Temporal.ZonedDateTime`
+- `Temporal.Duration`
+- `Temporal.Instant`
+- `Temporal.PlainYearMonth`
+- `Temporal.PlainMonthDay`
+
+But to be honest, this post isn't meant to give an overview of Temporal
+and its general API. If Temporal is news to you and you are interested
+in learning more, feel free to check out the phenomenal [MDN
+documentation][mdn-temporal].
+
+## Back on track
+
+Being a JavaScript engine / interpreter, implementing the ECMAScript
+specification is our raison d'être. One of our primary goals is to
+implement the specification to high degree of conformance. This is even
+more so important because Temporal represents roughly 7-8% of the
+current conformance test suite (~4000 of the ~50,000 tests).
+
+When the PR of the first prototype of Temporal for Boa was submitted a
+few things became evident.
+
+1. Date/Time is a complicated beast (duh)
+2. There's room for optimization and improvement
+3. This would be handy to have in Rust
+
+So after the prototype was merged, the prototype was pulled out of Boa's
+internal builtins and externalized into its own crate,
+[`temporal_rs`][temporal-rs-repo], which then first landed behind an
+experimental flag in Boa v0.18.
+
+After over a year and a half of development, Boa now sits at about 90%
+overall conformance for Temporal and growing with the entire
+implementation backed by `temporal_rs`.
+
+For its part, `temporal_rs` is shaping up to be a proper Rust date/time
+library that can be used to implement temporal in a JavaScript engine or
+for general usage purposes.
+
+So let's take a look at Temporal: it's JavaScript API, it's Rust API in
+`temporal_rs`, and how `temporal_rs` supports implementing the
+specification.
+
+## Important core differences
+
+Let's briefly talk about JavaScript values (`JsValue`). This is
```

Comment:
```suggestion
Firstly, JavaScript values (`JsValue`). This is
```

Reply:
Not sure this makes much sense. We'd have "Firstly, JavaScript values" ... maybe: "Firstly, let's briefly talk about JavaScript values".

Reply:
"First, we need to talk about JavaScript values (`JsValue`) for a bit." maybe?